### ISSUE TITLE
SPIR-V optimizer still breaks SMAA shaders

### ISSUE THREAD

**Issue by turol** [KhronosGroup/SPIRV-Tools#1388]
Even after #1385 was fixed the optimizer still breaks the shader.
[Good shader](https://gist.github.com/turol/5527f957a0c602c14a7f864ce061f442)
[Bad shader](https://gist.github.com/turol/054fdfdd966a03ba09163eda1f2ba586)
Please look at #1385 for other details.

**Comment by turol** [comment]
On further testing it looks like the Vulkan version is not affected. There are only tiny differences between OpenGL and Vulkan shader. Some constants are negated and there's at least one case of double negation which caused trouble with SPIRV-cross before. These differences are caused by the different Y coordinate conventions. Some arithmetic folding optimization is the likely culprit.

**Comment by alan-baker** [comment]
I've taken an initial look and I have yet to spot the problem. I will take a more in depth look soon. Given that only OpenGL shows a problem, it is possible these optimizations are exposing a bug elsewhere in your toolchain.

**Comment by turol** [comment]
The only other place where the bug could be would be SPIRV-cross. Unfortunately it wouldn't be easy to remove that.

**Comment by alan-baker** [comment]
Can you provide me the SPIRV-cross command you use on the SPIR-V? If I can't find the problem in the next round, I may take a look at what SPIRV-cross generates.

**Comment by turol** [comment]
It's not run as a separate command, it's part of the shader generation and caching system of smaaDemo. 

I tried with Vulkan using OpenGL -style shaders. The rendering was obviously wrong but not affected by optimization anymore. It looks more likely that SPIRV-cross is the culprit.

**Comment by turol** [comment]
Running standalone spirv-cross on the shaders doesn't produce exactly the right GLSL since the code changes the resource bindings by removing the set indices before converting to GLSL. However there should be no functional changes.

The actual shaders as seen by apitrace:

[Good GLSL shader](https://gist.github.com/turol/d5a9896ab9f887eb3cefd6251bec0386)
[Bad GLSL shader](https://gist.github.com/turol/341070b7197c2e92fb7c5bcdcf8cc593)

**Comment by dneto0** [comment]
Thanks for getting narrowing down to root cause.

**Comment by dneto0** [comment]
Closing in favour of https://github.com/KhronosGroup/SPIRV-Cross/issues/494
